### PR TITLE
doc: add a new document on distributed cephfs metadata cache

### PR DESCRIPTION
**Pull request by jtlayton**
Signed-off-by: Jeff Layton <jlayton@redhat.com>

### CODE REVIEW

**Comment by jtlayton** [comment]
Probably not ready for merge yet, but we can discuss.

**Comment by batrick** [comment]
jenkins render docs

File: doc/cephfs/mdcache.rst [hunk 1]
```diff
@@ -0,0 +1,80 @@
+=================================
+CephFS Distributed Metadata Cache
+=================================
+While the data for inodes in a CephFS filesystem is stored in RADOS and
```
Comment (batrick) [inline doc/cephfs/mdcache.rst]: nit: filesystem -> file system

File: doc/cephfs/mdcache.rst [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+client a certain set of caps for the inode, allowing it to perform
+subsequent requests without consulting the MDS.
+
+Clients can also request caps directly from the MDS, which is necessary
+in order to read or write file data.
+
+MDS Inode Locker
+----------------
+When an MDS wants to read or change information about an inode, it must
+gather the appropriate locks for it. If there are outstanding caps that
```
Comment (batrick) [inline doc/cephfs/mdcache.rst]: Good to mention that the locks are distributed and multi-faceted. The MDS cluster may have a series of different types of locks on the given inode and each MDS may have disjoint sets of locks.

@ukernel anything you want to add?

File: doc/cephfs/mdcache.rst [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+would conflict with these locks, then they must be revoked before the
+lock can be acquired. Once the competing caps are returned to the MDS,
+then it can get the locks and do the operation.
+
+Multiple MDS'
```
Comment (batrick) [inline doc/cephfs/mdcache.rst]: trailing `'`

File: doc/cephfs/mdcache.rst [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+On a filesystem served by multiple MDS', the metadata cache is also
+distributed among the MDS' in the cluster. For every inode, at any given
+time, only one MDS in the cluster is considered **authoritative**. Any
+requests to change that inode must be done by the authoritative MDS,
+though non-authoritative MDS can forward requests to the authoritative
+one.
+
+Non-auth MDS' can also obtain read locks that prevent the auth MDS from
+changing the data until the lock is dropped, so that they can serve
+inode info to the clients. [1][2]
```
Comment (batrick) [inline doc/cephfs/mdcache.rst]: I'd prefer to just merge this discussion of locks with the previous section and re-title it as "Distributed Locks in an MDS Cluster" or similar. Eventually we want to get to a point where we don't belabor the difference in behavior between 1 active MDS vs. N active MDS. 1 active MDS should just be a special case in the code (and it's not really).

File: doc/cephfs/mdcache.rst [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+The auth MDS for an inode can change over time as well. The MDS' will
+actively balance responsibility for the inode cache amongst
+themselves, but this can be overriden by **pinning** certain inodes to a
```
Comment (batrick) [inline doc/cephfs/mdcache.rst]: `s/inodes/subtrees/`

File: doc/cephfs/mdcache.rst [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+single MDS.
+
+[1]: Will they also grant shared/read caps under the aegis of a
+readlock, or do only auth MDS grant caps?
+
+[2]: Can a client return caps to a different MDS than the one that
+granted it?
```
Comment (batrick) [inline doc/cephfs/mdcache.rst]: @ukernel ?

**Comment by jtlayton** [comment]
Some updates, but I'm still leaving the DNM tag in place for now.

File: doc/cephfs/mdcache.rst [hunk 7]
```diff
@@ -0,0 +1,82 @@
+=================================
+CephFS Distributed Metadata Cache
+=================================
+While the data for inodes in a Ceph file system is stored in RADOS and
+accessed by the clients directly, inode metadata and directory
+information is managed by the Ceph metadata server (MDS).
```
Comment (fullerdj) [inline doc/cephfs/mdcache.rst]: I'd suggest mentioning that the metadata is also stored in RADOS. I find that people new to Ceph often assume it has a completely separate metadata store (like Lustre, for example).

File: doc/cephfs/mdcache.rst [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+**revoked**. When a capability is revoked, the client is responsible for
+returning it as soon as it is able. Clients that fail to do so in a
+timely fashion may end up **blacklisted** and unable to communicate with
+the cluster.
+
+Since the cache is distributed, the MDS must take great care to ensure
+that no client holds capabilities that may conflict with other clients'
+capabilities, or operations that it does itself.
+
+This allows cephfs clients to rely on much greater cache coherence than
```
Comment (rishabh-d-dave) [inline doc/cephfs/mdcache.rst]: Perhaps, merge this paragraph with the previous one? The separation makes me look at it again.

**Comment by ajarr** [comment]
looks great. waiting on @ukernel's inputs.

**Comment by jtlayton** [comment]
jenkins render docs

**Comment by jtlayton** [comment]
Removed the two questions I had in there. As best I can tell, I don't think non-auth MDS's will ever grant caps to clients. None of that is necessary in order to merge this though.

I'm going to go ahead and remove the DNM tag here. I think this is fine as-is.

I do not have any diagrams for this, but I think it'd be difficult to do one that clearly demonstrates the flow of caps between clients and MDS's.

What would be really cool is something animated: maybe 2 clients and an MDS and show the caps being granted and revoked in response to actions by the other clients. I'm not prepared to commit the time that it'd take to do that for this though, so I'll leave that for now.

**Pull request merged**